### PR TITLE
Web storage

### PR DESCRIPTION
**Pull request by MihlaliMgwebi**
Pulled from local storage if fetch unsuccessful.

### CODE REVIEW

File: src/services/dom.service.ts [hunk 1]
```diff
@@ -90,10 +90,18 @@ function createFlightDetailsCard(flight: IFlight) {
 }
 
 export function showSpinner() {
-    document.getElementById("app-main__spinner")?.classList.remove("hidden");
+    const spinner = document.getElementById("app-main__spinner")
```
Comment (sechaba-mache) [inline src/services/dom.service.ts]: Please add types

File: src/services/dom.service.ts [hunk 2]
```diff
@@ -90,10 +90,18 @@ function createFlightDetailsCard(flight: IFlight) {
...
+    spinner?.classList.remove("hidden");
+    spinner?.classList.add("flex", "justify-center", "items-center");
 }
 
 export function hideSpinner() {
-    document.getElementById("app-main__spinner")?.classList.add("hidden");
+    const spinner = document.getElementById("app-main__spinner")
```
Comment (sechaba-mache) [inline src/services/dom.service.ts]: Please add types

File: src/services/dom.service.ts [hunk 3]
```diff
@@ -90,10 +90,18 @@ function createFlightDetailsCard(flight: IFlight) {
...
+    spinner?.classList.add("hidden");
+    spinner?.classList.remove("flex", "justify-center", "items-center");
 }
 
+export function showLeafletMap() {
+    const leafletMapContainer = document.getElementById('app-main__map')
```
Comment (sechaba-mache) [inline src/services/dom.service.ts]: Please add types

File: src/services/api.service.ts [hunk 4]
```diff
@@ -47,14 +50,26 @@ export function pollFirst20FlightDetails(timeInMilliseconds: number): Observable
...
+                        localStorage.setItem(localStorageKey, JSON.stringify(flights));
                         return flights;
                     }
-                    else {
-                        return {
-                            flights: []
+                    else { // if error mesage
+                        if (cachedResponse) {
+                            try {
+                                const cachedFlights = JSON.parse(cachedResponse);
```
Comment (sechaba-mache) [inline src/services/api.service.ts]: Please add types

File: src/services/api.service.ts [hunk 5]
```diff
@@ -1,18 +1,21 @@
...
 const BASE_URL = "https://opensky-network.org/api/states/all?extended=1&time=";
 
 function isIFlightAPIResponse(result: IFlightAPIResponse | { error: boolean; message: string; }): result is IFlightAPIResponse {
     return Object.keys(result).includes("states")
 }
 
 export function pollFirst20FlightDetails(timeInMilliseconds: number): Observable<IFlights> {
-    // const url = 'src/services/data.json';
+    const localStorageKey = `pollFirst20FlightDetails:${timeInMilliseconds}`;
+    const cachedResponse = localStorage.getItem(localStorageKey);
```
Comment (sechaba-mache) [inline src/services/api.service.ts]: Please add types

File: src/main.ts [hunk 6]
```diff
@@ -85,13 +83,22 @@ allFlightsStream$.subscribe((allFlights) => {
...
+    // Show map and Add new flight summary and details containers
     if (allFlights && allFlights.flights) {
+      showLeafletMap()
       allFlights.flights.forEach((flight) => {
         const newFlightSummaryAndDetailsContainer = createOneFlightSummaryAndDetailsContainer(flight);
         allFlightSummaryAndDetailsContainer.appendChild(newFlightSummaryAndDetailsContainer);
       });
     }
+    else {
+      const errorHandlingMessage = document.createElement("p")
```
Comment (sechaba-mache) [inline src/main.ts]: Please add types

**Review by sechaba-mache — APPROVED** [review]

File: src/services/dom.service.ts (hunk 1 above)
Comment (MihlaliMgwebi) [reply src/services/dom.service.ts]: Thank you for you advice. Explicitly typed as requested.

**Pull request merged**